// postgrejs/cjs/data-types/oid-type.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VectorOidType = exports.ArrayOidType = exports.OidType = void 0;
const constants_js_1 = require("../constants.js");
const fast_parseint_js_1 = require("../util/fast-parseint.js");
exports.OidType = {
    name: 'oid',
    oid: constants_js_1.DataTypeOIDs.oid,
    jsType: 'number',
    parseBinary(v) {
        return v.readUInt32BE(0);
    },
    encodeBinary(buf, v) {
        buf.writeUInt32BE((0, fast_parseint_js_1.fastParseInt)(v));
    },
    parseText: fast_parseint_js_1.fastParseInt,
    isType(v) {
        return typeof v === 'number' && Number.isInteger(v) && !!constants_js_1.DataTypeNames[v];
    },
};
exports.ArrayOidType = {
    ...exports.OidType,
    name: '_oid',
    oid: constants_js_1.DataTypeOIDs._oid,
    elementsOID: constants_js_1.DataTypeOIDs.oid,
};
exports.VectorOidType = {
    ...exports.OidType,
    name: '_oidvector',
    oid: constants_js_1.DataTypeOIDs.oidvector,
    elementsOID: constants_js_1.DataTypeOIDs.oid,
};
