// postgrejs/cjs/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataTypeNames = exports.DataTypeOIDs = exports.ConnectionState = exports.DEFAULT_COLUMN_FORMAT = exports.DataFormat = void 0;
const protocol_js_1 = require("./protocol/protocol.js");
var DataFormat = protocol_js_1.Protocol.DataFormat;
exports.DataFormat = DataFormat;
// @ts-ignore
Symbol.asyncDispose ?? (Symbol.asyncDispose = Symbol('Symbol.asyncDispose'));
exports.DEFAULT_COLUMN_FORMAT = DataFormat.binary;
var ConnectionState;
(function (ConnectionState) {
    ConnectionState[ConnectionState["CLOSED"] = 0] = "CLOSED";
    ConnectionState[ConnectionState["CONNECTING"] = 1] = "CONNECTING";
    ConnectionState[ConnectionState["AUTHORIZING"] = 3] = "AUTHORIZING";
    ConnectionState[ConnectionState["READY"] = 2] = "READY";
    ConnectionState[ConnectionState["CLOSING"] = 10] = "CLOSING";
})(ConnectionState || (exports.ConnectionState = ConnectionState = {}));
exports.DataTypeOIDs = {
    bool: 16,
    bytea: 17,
    char: 18,
    name: 19,
    int8: 20,
    int2: 21,
    int2vector: 22,
    int4: 23,
    regproc: 24,
    text: 25,
    oid: 26,
    tid: 27,
    xid: 28,
    cid: 29,
    oidvector: 30,
    json: 114,
    xml: 142,
    point: 600,
    lseg: 601,
    path: 602,
    box: 603,
    polygon: 604,
    line: 628,
    cidr: 650,
    float4: 700,
    float8: 701,
    unknown: 705,
    circle: 718,
    macaddr8: 774,
    money: 790,
    macaddr: 829,
    inet: 869,
    bpchar: 1042,
    varchar: 1043,
    date: 1082,
    time: 1083,
    timestamp: 1114,
    timestamptz: 1184,
    interval: 1186,
    timetz: 1266,
    bit: 1560,
    varbit: 1562,
    numeric: 1700,
    refcursor: 1790,
    regprocedure: 2202,
    regoper: 2203,
    regoperator: 2204,
    regclass: 2205,
    regtype: 2206,
    record: 2249,
    cstring: 2275,
    any: 2276,
    anyarray: 2277,
    void: 2278,
    trigger: 2279,
    // eslint-disable-next-line camelcase
    language_handler: 2280,
    internal: 2281,
    anyelement: 2283,
    anynonarray: 2776,
    uuid: 2950,
    jsonb: 3802,
    anyrange: 3831,
    int4range: 3904,
    numrange: 3906,
    tsrange: 3908,
    rstzrange: 3910,
    daterange: 3912,
    int8range: 3926,
    _xml: 143,
    _json: 199,
    _xid8: 271,
    _line: 629,
    _cidr: 651,
    _circle: 719,
    _macaddr8: 775,
    _money: 791,
    _bool: 1000,
    _bytea: 1001,
    _char: 1002,
    _name: 1003,
    _int2: 1005,
    _int2vector: 1006,
    _int4: 1007,
    _regproc: 1008,
    _text: 1009,
    _tid: 1010,
    _xid: 1011,
    _cid: 1012,
    _oidvector: 1013,
    _bpchar: 1014,
    _varchar: 1015,
    _int8: 1016,
    _point: 1017,
    _lseg: 1018,
    _path: 1019,
    _box: 1020,
    _float4: 1021,
    _float8: 1022,
    _polygon: 1027,
    _oid: 1028,
    _macaddr: 1040,
    _inet: 1041,
    _timestamp: 1115,
    _date: 1182,
    _time: 1183,
    _timestamptz: 1185,
    _interval: 1187,
    _numeric: 1231,
    _cstring: 1263,
    _timetz: 1270,
    _bit: 1561,
    _varbit: 1563,
    _uuid: 2951,
    _jsonb: 3807,
};
exports.DataTypeNames = {
    [exports.DataTypeOIDs.bool]: 'bool',
    [exports.DataTypeOIDs.bytea]: 'bytea',
    [exports.DataTypeOIDs.char]: 'char',
    [exports.DataTypeOIDs.name]: 'name',
    [exports.DataTypeOIDs.int8]: 'int8',
    [exports.DataTypeOIDs.int2]: 'int2',
    [exports.DataTypeOIDs.int2vector]: 'int2vector',
    [exports.DataTypeOIDs.int4]: 'int4',
    [exports.DataTypeOIDs.regproc]: 'regproc',
    [exports.DataTypeOIDs.text]: 'text',
    [exports.DataTypeOIDs.oid]: 'oid',
    [exports.DataTypeOIDs.tid]: 'tid',
    [exports.DataTypeOIDs.xid]: 'xid',
    [exports.DataTypeOIDs.cid]: 'cid',
    [exports.DataTypeOIDs.oidvector]: 'oidvector',
    [exports.DataTypeOIDs.json]: 'json',
    [exports.DataTypeOIDs.xml]: 'xml',
    [exports.DataTypeOIDs.point]: 'point',
    [exports.DataTypeOIDs.lseg]: 'lseg',
    [exports.DataTypeOIDs.path]: 'path',
    [exports.DataTypeOIDs.box]: 'box',
    [exports.DataTypeOIDs.polygon]: 'polygon',
    [exports.DataTypeOIDs.line]: 'line',
    [exports.DataTypeOIDs.cidr]: 'cidr',
    [exports.DataTypeOIDs.float4]: 'float4',
    [exports.DataTypeOIDs.float8]: 'float8',
    [exports.DataTypeOIDs.unknown]: 'unknown',
    [exports.DataTypeOIDs.circle]: 'circle',
    [exports.DataTypeOIDs.macaddr8]: 'macaddr8',
    [exports.DataTypeOIDs.money]: 'money',
    [exports.DataTypeOIDs.macaddr]: 'macaddr',
    [exports.DataTypeOIDs.inet]: 'inet',
    [exports.DataTypeOIDs.bpchar]: 'bpchar',
    [exports.DataTypeOIDs.varchar]: 'varchar',
    [exports.DataTypeOIDs.date]: 'date',
    [exports.DataTypeOIDs.time]: 'time',
    [exports.DataTypeOIDs.timestamp]: 'timestamp',
    [exports.DataTypeOIDs.timestamptz]: 'timestamptz',
    [exports.DataTypeOIDs.interval]: 'interval',
    [exports.DataTypeOIDs.timetz]: 'timetz',
    [exports.DataTypeOIDs.bit]: 'bit',
    [exports.DataTypeOIDs.varbit]: 'varbit',
    [exports.DataTypeOIDs.numeric]: 'numeric',
    [exports.DataTypeOIDs.refcursor]: 'refcursor',
    [exports.DataTypeOIDs.regprocedure]: 'regprocedure',
    [exports.DataTypeOIDs.regoper]: 'regoper',
    [exports.DataTypeOIDs.regoperator]: 'regoperator',
    [exports.DataTypeOIDs.regclass]: 'regclass',
    [exports.DataTypeOIDs.regtype]: 'regtype',
    [exports.DataTypeOIDs.record]: 'record',
    [exports.DataTypeOIDs.cstring]: 'cstring',
    [exports.DataTypeOIDs.any]: 'any',
    [exports.DataTypeOIDs.anyarray]: 'anyarray',
    [exports.DataTypeOIDs.void]: 'void',
    [exports.DataTypeOIDs.trigger]: 'trigger',
    [exports.DataTypeOIDs.language_handler]: 'language_handler',
    [exports.DataTypeOIDs.internal]: 'internal',
    [exports.DataTypeOIDs.anyelement]: 'anyelement',
    [exports.DataTypeOIDs.anynonarray]: 'anynonarray',
    [exports.DataTypeOIDs.uuid]: 'uuid',
    [exports.DataTypeOIDs.jsonb]: 'jsonb',
    [exports.DataTypeOIDs.anyrange]: 'anyrange',
    [exports.DataTypeOIDs.int4range]: 'int4range',
    [exports.DataTypeOIDs.numrange]: 'numrange',
    [exports.DataTypeOIDs.tsrange]: 'tsrange',
    [exports.DataTypeOIDs.rstzrange]: 'rstzrange',
    [exports.DataTypeOIDs.daterange]: 'daterange',
    [exports.DataTypeOIDs.int8range]: 'int8range',
    [exports.DataTypeOIDs._xml]: '_xml',
    [exports.DataTypeOIDs._json]: '_json',
    [exports.DataTypeOIDs._xid8]: '_xid8',
    [exports.DataTypeOIDs._line]: '_line',
    [exports.DataTypeOIDs._cidr]: '_cidr',
    [exports.DataTypeOIDs._circle]: '_circle',
    [exports.DataTypeOIDs._macaddr8]: '_macaddr8',
    [exports.DataTypeOIDs._money]: '_money',
    [exports.DataTypeOIDs._bool]: '_bool',
    [exports.DataTypeOIDs._bytea]: '_bytea',
    [exports.DataTypeOIDs._char]: '_char',
    [exports.DataTypeOIDs._name]: '_name',
    [exports.DataTypeOIDs._int2]: '_int2',
    [exports.DataTypeOIDs._int2vector]: '_int2vector',
    [exports.DataTypeOIDs._int4]: '_int4',
    [exports.DataTypeOIDs._regproc]: '_regproc',
    [exports.DataTypeOIDs._text]: '_text',
    [exports.DataTypeOIDs._tid]: '_tid',
    [exports.DataTypeOIDs._xid]: '_xid',
    [exports.DataTypeOIDs._cid]: '_cid',
    [exports.DataTypeOIDs._oidvector]: '_oidvector',
    [exports.DataTypeOIDs._bpchar]: '_bpchar',
    [exports.DataTypeOIDs._varchar]: '_varchar',
    [exports.DataTypeOIDs._int8]: '_int8',
    [exports.DataTypeOIDs._point]: '_point',
    [exports.DataTypeOIDs._lseg]: '_lseg',
    [exports.DataTypeOIDs._path]: '_path',
    [exports.DataTypeOIDs._box]: '_box',
    [exports.DataTypeOIDs._float4]: '_float4',
    [exports.DataTypeOIDs._float8]: '_float8',
    [exports.DataTypeOIDs._polygon]: '_polygon',
    [exports.DataTypeOIDs._oid]: '_oid',
    [exports.DataTypeOIDs._macaddr]: '_macaddr',
    [exports.DataTypeOIDs._inet]: '_inet',
    [exports.DataTypeOIDs._timestamp]: '_timestamp',
    [exports.DataTypeOIDs._date]: '_date',
    [exports.DataTypeOIDs._time]: '_time',
    [exports.DataTypeOIDs._timestamptz]: '_timestamptz',
    [exports.DataTypeOIDs._interval]: '_interval',
    [exports.DataTypeOIDs._numeric]: '_numeric',
    [exports.DataTypeOIDs._cstring]: '_cstring',
    [exports.DataTypeOIDs._timetz]: '_timetz',
    [exports.DataTypeOIDs._bit]: '_bit',
    [exports.DataTypeOIDs._varbit]: '_varbit',
    [exports.DataTypeOIDs._uuid]: '_uuid',
    [exports.DataTypeOIDs._jsonb]: '_jsonb',
};
